### ISSUE TITLE
would you report mIoU on pascal voc dataset?

### ISSUE THREAD

**Issue by jianlong-yuan** [holyseven/PSPNet-TF-Reproduce#1]

**Comment by holyseven** [comment]
Training with 12031 (10582 + 1449, train and validation set) images, [L2-SP regularization](https://github.com/holyseven/TransferLearningClassification), and original ResNet-101, I got [79.8 scores mIoU ](http://host.robots.ox.ac.uk:8080/anonymous/NAAVTI.html)(compared to 82.6 in the paper) on the test set. Not sure if changing the structure of network will help 3 points.

**Comment by yjl9122** [comment]
do you train with trainval date?

**Comment by jianlong-yuan** [comment]
@holyseven do you use trainval data set?

**Comment by holyseven** [comment]
yes

**Comment by jianlong-yuan** [comment]
@holyseven How many points improve with L2-SP?

**Comment by holyseven** [comment]
Training with only train set (10582 images), tested on val set (1449), L2 got 78.3, L2-SP 79.9.

**Comment by yjl9122** [comment]
@holyseven Maybe, test with sliding window will be better.

I found there is so many modes, so which mode you chose? If weight_decay_mode=1?

**Comment by yjl9122** [comment]
@holyseven Maybe, test with sliding window will be better.

I found there is so many modes, so which mode you chose? If weight_decay_mode=1?
Could you tell us how to set your parameters related to L2-SP?

**Comment by jianlong-yuan** [comment]
@holyseven Maybe, test with sliding window will be better.

I found there is so many modes, so which mode you chose? If weight_decay_mode=1?
Could you tell us how to set your parameters related to L2-SP on cityscapes and Pascal VOC?

**Comment by holyseven** [comment]
Detailed in the [paper](https://arxiv.org/abs/1802.01483). 

Simply, the reference of regularization for the parameters is the pre-trained model. Here, the pre-trained model I used is ResNet-V1-101, proposed by [Tensorflow/slim](https://github.com/tensorflow/models/tree/master/research/slim#pre-trained-models), converted from the original caffe format. Examples of scripts for fine-tuning on Cityscapes and Pascal VOC are proposed in the [README.md](https://github.com/holyseven/PSPNet-TF-Reproduce#training).

**Comment by fanq15** [comment]
When I train with only train set (10582 images), tested on val set (1449), L2 got 76.6, L2-SP 78.4.

**Comment by PkuRainBow** [comment]
@fanq15 So you means that L2-SP can improve 2 points compared with the baseline method?

**Comment by PkuRainBow** [comment]
@fanq15 Have you tried this method with Pytorch?  It seems really easy to use.

**Comment by John1231983** [comment]
Hi all, anyone who success to implement L2-SP in pytorch? I would like to test the efficient of L2-SP using pytorch ? Thanks

**Comment by holyseven** [comment]
I wrote some codes in pytorch to implement L2-SP (that hasn't been tested yet):

``` python
import torch
import torchvision

# download the pretrained weights from https://download.pytorch.org/models/resnet101-5d3b4d8f.pth
pretrained_weights = torch.load('./resnet101-5d3b4d8f.pth')
print pretrained_weights.keys()

# a = pretrained_weights['layer1.0.bn1.running_mean']
# print type(a), type(a.data)  # this is a `Tensor`

resnet101_model = torchvision.models.resnet101(num_classes=120)

existing_l2_reg = 0.0
new_l2_reg = 0.0
for name, w in resnet101_model.named_parameters():
    if 'weight' not in name:  # I don't know if that is true: I was told that Facebook regularized biases too.
        continue
    if 'downsample.1' in name:  # another bias
        continue
    if 'bn' in name:  # bn parameters
        continue

    if 'fc' in name:
        new_l2_reg += torch.pow(w, 2).sum()
        print 'L2:', name, w.size()
    else:
        w0 = pretrained_weights[name].data
        # if I didn't misunderstand,
        # pretrained_weights[name] is a `Parameter`, which can be trained;
        # pretrained_weights[name].data is a `Tensor`, which is considered as a constant.
        # So here we just want to use w0 as constant and train w.
        # print type(w0), type(w),
        print 'L2-SP:', name, w.size()
        existing_l2_reg += torch.pow(w-w0, 2).sum()

# l2_reg = existing_l2_reg * rate1 + new_l2_reg * rate2
# totol_loss += l2_reg
# ...
```

I have no plans to continue with pytorch at this moment. If you could try L2-SP with some experiments in pytorch (classification or segmentation, or other transfer learning applications), thanks in advance for sharing.

**UPDATE**
w.norm(2) => torch.pow(w, 2).sum()
Still, I haven't tested this implementation.
